modules/channel-vonage/src/backend/client.ts

```diff
@@ -208,167 +213,50 @@ export class VonageClient {
-        } else if (button.type === 'say_something') {
-          options.push({ label: title, value: button.text as string, type: 'say_something' })
-        }
-      }
-
-      if (picture) {
-        await this.sendImage(event, { url: picture, type: 'image' })
-      }
-
-      await this.sendOptions(event, body, options)
     }
-  }
 
-  // TODO: add support for WhatsApp Templates instead of using down rendering
-  private async sendOptions(event: sdk.IO.Event, text: string, options: MessageOption[]) {
-    if (options.length) {
-      text = `${text}\n\n${options.map(({ label }, idx) => `*(${idx + 1})* ${label}`).join('\n')}`
+    for (const sender of this.senders) {
+      if (sender.handles(context)) {
+        await sender.send(context)
+      }
     }
 
-    await this.kvs.set(this.getKvsKey(event.target, event.threadId), options, undefined, '10m')
+    await this.bp.experimental.messages
+      .forBot(this.botId)
+      .create(event.threadId, event.payload, undefined, event.id, event.incomingEventId)
 
-    await this.sendMessage(event, { type: 'text', text })
+    next(undefined, false)
   }
 
-  private async sendMessage(event: sdk.IO.Event, content: VonageChannelContent, whatsapp?: ChannelWhatsApp) {
-    const message: ChannelMessage = {
-      content,
-      whatsapp
+  async handleIncomingEvent(event: sdk.IO.IncomingEvent, next: sdk.IO.MiddlewareNextCallback) {
+    const index = Number(event.payload.text)
+    if (!index) {
+      return
     }
 
-    debugOutgoing('Sending message', JSON.stringify(message, null, 2))
-
-    await new Promise(resolve => {
-      this.vonage.channel.send(
-        { type: SUPPORTED_CHANNEL_TYPE, number: event.target },
-        {
-          type: SUPPORTED_CHANNEL_TYPE,
-          number: this.config.botPhoneNumber
-        },
-        message,
-        (err, data) => {
-          if (err) {
-            // fixes typings
-            const errBody: MessageApiError = (err as any).body
-            let reasons: string = ''
-            if (errBody) {
-              if (errBody.invalid_parameters) {
-                for (const param of errBody.invalid_parameters) {
-                  reasons += `${param.reason}: ${param.name}; `
-                }
-              }
-
-              this.logger.error(`${errBody.title}: ${errBody.detail} ${reasons}${errBody.type}`)
-            } else if ((<any>err).statusCode === '429') {
-              this.logger.error('HTTPError (429): Too Many Requests')
-            } else {
-              this.logger.error('UnknownError', err)
-            }
-          } else {
-            resolve(data)
-          }
-        }
-      )
-    })
-
-    // Sandbox API is limited to one call per second. Wait one second between calls.
-    if (this.config.useTestingApi) {
-      await Promise.delay(1000)
+    const mappedPayload = (await this.handleIndexReponse(index - 1, event.target, event.threadId)) ?? event.payload
+    if (!mappedPayload.text) {
```

Comment:
Maybe just validate that mappedPayload is defined? Would it be possible for the function to return an object that contains an empty text property?

Reply:
`Say something` and `Postback` both have a text value but `OpenURL` doesn't. Choices and dropdown also all set a text value. So we swallow the event here when the user responsds by selection a url option for some reason.